src/converter/one/heatmap/Heatmap.java

```diff
@@ -395,35 +387,53 @@ private static class State {
 
         // Maps stack trace ID to prototype ID in stackTracesRemap
         final DictionaryInt stackTracesCache = new DictionaryInt();
-        final MethodCache methodsCache;
+        final Map<MethodKey, Integer> methodCache = new HashMap<>();
+        final Index<Method> methods = new Index<>(Method.class, Method.EMPTY);
+        final Index<String> symbolTable = new Index<>(String.class, "");
+        final Arguments args;
+
+        final JfrConverter converter;
 
         // reusable array to (temporary) store (potentially) new stack trace
         int[] cachedStackTrace = new int[4096];
 
-        State(JfrConverter converter, long blockDurationMs) {
+        State(JfrConverter converter, long blockDurationMs, Arguments args) {
             sampleList = new SampleList(blockDurationMs);
-            methodsCache = new MethodCache(converter);
+            this.converter = converter;
+            this.args = args;
         }
 
-        public void addEvent(int stackTraceId, int extra, byte type, long timeMs) {
+        public void addEvent(int stackTraceId, int threadId, int classId, byte type, long timeMs) {
             if (sampleList.getRecordsCount() >= LIMIT) {
                 return;
             }
 
             int prototypeId = stackTracesCache.get(stackTraceId);
-            if (extra == 0) {
+            if (classId == 0 && !args.threads) {
                 sampleList.add(prototypeId, timeMs);
                 return;
             }
 
             int[] prototype = stackTracesRemap.get(prototypeId);
-            int stackSize = prototype.length + 1;
+            int stackSize = prototype.length + (args.threads ? 1 : 0) + (classId != 0 ? 1 : 0);
             if (cachedStackTrace.length < stackSize) {
                 cachedStackTrace = new int[stackSize * 2];
             }
 
-            System.arraycopy(prototype, 0, cachedStackTrace, 0, prototype.length);
-            cachedStackTrace[prototype.length] = methodsCache.indexForClass(extra, type);
+            if (args.threads) {
+                MethodKey key = new MethodKey(threadId, -1, Frame.TYPE_NATIVE, true);
+                Integer threadFrameIndex = getMethodIndex(key, () -> createThreadMethod(key, threadId));
+                cachedStackTrace[0] = threadFrameIndex;
+            }
+
+            System.arraycopy(prototype, 0, cachedStackTrace, args.threads ? 1 : 0, prototype.length);
+
+            if (classId != 0) {
+                MethodKey key = new MethodKey((long) classId << 32, -1, type, false);
+                Integer threadFrameIndex = getMethodIndex(key, () -> createClassMethod(key, classId));
```

Comment:
It's not `threadFrameIndex`. Anyhow, I'd remove a temporary variable altogether. In the above `if` too.

Reply:
Right, thanks: 4804b5dbca15adf5d01591a09f3e8c263d756a9f